docs/04-reference/winglang-spec.md

```diff
@@ -172,45 +172,389 @@ inflight (arg1: <type1>, arg2: <type2>, ...) => <type>
-> let data: Struct = await fetch_some_data_with_jsii();
-> print(data[0]["prop-1"].id);
-> print(data.name);
-> ```
-  
-<details><summary>Equivalent TypeScript Code</summary>
+Let's look at a quick example:
 
-> ```TS
-> const data: any = Object.freeze(await (async () => {
->   try {
->     // deep copy all enumerable properties of the host object
->     return JSON.parse(JSON.stringify(await fetch_some_data_with_jsii()));
->   } catch(err) {
->     throw new Error(`Failed to parse JSON from any: ${err.message}`);
->   }
-> })());
-> console.log(data[0]["prop-1"].id);
-> console.log(data.name);
-> ```
+```js
+struct Employee { 
+  id: str;
+  name: str;
+}
 
-</details>
+let response = http_get("/employees"); // returns something like { "items": [ { "id": "12234", "name": "bob" }, ... ] }
+let employees = Array<Employee>.from_json(response.items);
+
+for e in employees {
+  print("hello, ${e.name}, your employee id is ${id}");
+}
+```
+
+In the above example, the `http_get` function returns a `Json` object from the server that has a
+single field `items`, with a JSON array of JSON objects, each with an `id` and `name` fields.
+
+The expression `response.items` returns a `Json` array, and we use `Array<T>.from_json` to convert
+this array from `Json` to an `Array<Employee>`. Note that by default `from_json` will perform schema
+validation on the array and on each item (based on the declaration of the `Employee` struct).
+
+##### 1.1.4.1 Literals
+
+Literals can be defined using the `Json` type initializers:
+
+```js
+let json_string  = Json "hello";
+let json_number  = Json 123;
+let json_bool    = Json true;
+let json_array   = Json [ 1, 2, 3 ];
+let json_obj     = Json { boom: 123 };
+let json_mut_obj = MutJson {
+  hello: 123, 
+  world: [ 1, "cat", 3 ],       // <-- heterogenous array
+  "boom boom": { hello: 1233 }  // <-- non-symbolic key
+};
+```
+
+Every value within a `Json` array or object also has a type of `Json`.
+
+##### 1.1.4.2 JSON objects
+
+To access a field within an object, use the `.` notation:
+
+```js
+let boom: Json = json_obj.boom;
+```
+
+Trying to access a non-existent field will fail at runtime. For example:
+
+```js
+print(json_obj.boom.dude.world);
+// RUNTIME ERROR: Uncaught TypeError: Cannot read properties of undefined (reading 'world')
+```
+
+Like in JavaScript, it is also possible to access object fields using `[]`:
+
+```js
+let foo = j["my-field"].your_field["their-field"];
+```
+
+To obtain an array of all the keys within a JSON object use the `Json.keys(o)` method. 
+
+
+```js
+let j = { hello: 123, world: [ 1, 2, 3 ] };
+assert(Json.keys(j) == ["hello", "world"]);
+```
+
+To obtain an array of all the values, use `Json.values(o)`. To obtain an array of all key/value
+pairs use `Json.entries(o)` (P2):
+
+```js
+assert(Json.values(j).equals(Json [ 123, [ 1, 2, 3 ] ]));
+assert(Json.entries(j).equals(Json [
+  [ "hello", 123 ],
+  [ "world", [ 1, 2, 3 ] ]
+]));
```

Comment:
I would actually expect:
```suggestion
assert(Json.values(j).equals([ Json 123, Json [ 1, 2, 3 ] ]));
assert(Json.entries(j).equals([
  [ Json "hello", Json 123 ],
  [ Json "world", Json [ 1, 2, 3 ] ]
]));
```
TL;DR we can provide a more specific type to users -- `Array<Json>` instead of `Json` in the first example, and `Array<Array<Json>>` instead of `Json` in the second example.

Reply:
cool